### PR TITLE
BUG: Metadata propagation for groupby iterator

### LINKED ISSUE #37343: BUG: groupby __iter__ on pandas 1.1.x not propagating _metadata on DataFrame subclasses
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python

In [1]: import pandas as pd
   ...:
   ...: class SubclassedDataFrame2(pd.DataFrame):
   ...:
   ...:     # temporary properties
   ...:     _internal_names = pd.DataFrame._internal_names + ['internal_cache']
   ...:     _internal_names_set = set(_internal_names)
   ...:
   ...:     # normal properties
   ...:     _metadata = ['added_property']
   ...:
   ...:     @property
   ...:     def _constructor(self):
   ...:         return SubclassedDataFrame2
   ...:

In [2]: df = SubclassedDataFrame2({'A': [1, 2, 3], 'B': [1, 1, 2], 'C': [7, 8, 9]})
   ...: df.added_property = "hello"
   ...: for i, d in df.groupby("B"):
   ...:     assert d.added_property == "hello"
   ...:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-bf64fc505324> in <module>
      2 df.added_property = "hello"
      3 for i, d in df.groupby("B"):
----> 4     assert d.added_property == "hello"
      5

~\RMS-Python\lib\site-packages\pandas\core\generic.py in __getattr__(self, name)
   5133             or name in self._accessors
   5134         ):
-> 5135             return object.__getattribute__(self, name)
   5136         else:
   5137             if self._info_axis._can_hold_identifiers_and_holds_name(name):

AttributeError: 'SubclassedDataFrame2' object has no attribute 'added_property'


```

#### Problem description

Under pandas 1.0.x the above is working, on 1.1.x, the _metadata is not propagated to the split frames. I think this is a known issue with groupby and `__finalize__`, but this specific behavior with `__iter__` is a regression from 1.0.x behavior on DataFrame subclasses.

#### Expected Output

Assert statements above are passing, with metadata propagated to the split dataframe

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : db08276bc116c438d3fdee492026f8223584c477
python           : 3.7.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : AMD64 Family 23 Model 113 Stepping 0, AuthenticAMD
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.1.3
numpy            : 1.18.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : 0.29.17
pytest           : 5.4.1
hypothesis       : 5.36.1
sphinx           : 2.4.4
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.9.0
bottleneck       : 1.3.2
fsspec           : 0.8.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.1.3
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.16
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.49.0

</details>

### PR DESCRIPTION
**Pull request by Japanuspus**
Addresses part of #28283

This PR ensures that `__finalize__` is called for objects returned by iterator over `.groupby`-results.
Based on the discussion and benchmarks in PR #35688 (see https://github.com/pandas-dev/pandas/pull/35688#issuecomment-675367517), `__finalize__` is _not_ called for intermediate objects used by `.apply` (and maybe other code paths).

The implemented solution is the safe choice with regard to performance: An alternative solution would be to call `__finalize__` immediately after initialization (in `DataSplitter._chop()`).
The downside to doing so would be the performance hit due to the overhead of finalizing intermediate objects in `.apply` (and maybe other code paths).  In the benchmarks referenced above, this amounted to around 4%.
The upside of finalizing immediately after initialization would be reduced complexity and that it would allow `.apply` (and any other code paths using `DataSplitter._chop` to access metadata)..

- [x] closes #37343 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by Japanuspus** [comment]
@jorisvandenbossche You asked about this PR. Do you have any position on the performance question? Would you be able to review?

I will have a look at the failing tests.

**Comment by simonjayhawkins** [comment]
Thanks @Japanuspus for the PR.

> Addresses part of #28283

This PR may also fix #37343, see https://github.com/pandas-dev/pandas/issues/37343#issuecomment-718864907

If you could add a test for that issue, that'll be great. Also if you could keep the changes in this PR minimal as we would likely want to backport this.

**Comment by Japanuspus** [comment]
Thanks @simonjayhawkins! 

> This PR may also fix #37343, see [#37343 (comment)](https://github.com/pandas-dev/pandas/issues/37343#issuecomment-718864907)
> 
> If you could add a test for that issue, that'll be great. Also if you could keep the changes in this PR minimal as we would likely want to backport this.

Yes, this was exactly the issue that I saw in my own code as well after the 1.1 update. I will add a test. It appears I need to merge 2414c75bfe06064a95d8da4db5c8248e5aa0cdbb to get passing tests: will rebase and force push to keep things clean for backport.

**Comment by Japanuspus** [comment]
Still seeing unrelated test failures. Will try waiting for things to settle down once 1.1.4 is released.

File: pandas/tests/groupby/test_grouping.py [hunk 1]
```diff
@@ -963,3 +963,18 @@ def test_grouping_string_repr(self):
         result = gr.grouper.groupings[0].__repr__()
         expected = "Grouping(('A', 'a'))"
         assert result == expected
+
```
Comment (jreback) [inline pandas/tests/groupby/test_grouping.py:966]: hmm i think we have other metadata tests for groupby, so nothing broke?

**Review by jreback — CHANGES_REQUESTED** [review]
can you also update the metadata issue in the 1.2 whatsnew (just add the issue number on)

**Comment by jbrockmendel** [comment]
Does something similar need to be done for the cython path?

**Comment by Japanuspus** [comment]
> Does something similar need to be done for the cython path?

I believe not: The functional change is in `BaseGrouper.get_iterator` which does not appear to have a cython alternative path.

**Comment by simonjayhawkins** [comment]
@jreback are you OK with backporting this?

**Comment by jreback** [comment]
> @jreback are you OK with backporting this?

actually this is fine to backport. 

@Japanuspus can you update the whatsnew note to put in 1.1.5

File: doc/source/whatsnew/v1.2.0.rst [hunk 2]
```diff
@@ -558,6 +558,7 @@ Other
 - Bug in :meth:`Index.union` behaving differently depending on whether operand is a :class:`Index` or other list-like (:issue:`36384`)
 - Passing an array with 2 or more dimensions to the :class:`Series` constructor now raises the more specific ``ValueError``, from a bare ``Exception`` previously (:issue:`35744`)
 - Bug in ``accessor.DirNamesMixin``, where ``dir(obj)`` wouldn't show attributes defined on the instance (:issue:`37173`).
+- Bug in metadata propagation for ``groupby`` iterator (:issue:`37343`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:561]: move to 1.1.5

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm if you can move the note and merge master. ping on green-ish.

File: doc/source/whatsnew/v1.2.0.rst (hunk 2 above)
Comment (Japanuspus) [reply doc/source/whatsnew/v1.2.0.rst]: Done. Force pushed to drop this commit.

**Comment by simonjayhawkins** [comment]
> lgtm if you can move the note and merge master. ping on green-ish.

32 bit failure unrelated. (caused by recent merge to master of #36842)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Japanuspus

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Pull request merged**